### PR TITLE
Real SC unit test, with json format

### PR DESCRIPTION
**Pull request by shargon**
Still working on this, don't merge

### CODE REVIEW

**Comment by shargon** [comment]
@osmirnov we need to use `InteropService` with dummy storages (Dictionaries) without write all the logic again for unit testing

**Review by osmirnov — APPROVED** [review]

**Comment by shargon** [comment]
@osmirnov please review the json format. If you think that we need to rename some of them tell me :)

File: test/NeoSharp.VM.Test/Tests/PUSHM1_to_PUSH16.json [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+                    "state":
+                    {
+                        "state": "None",
+                        "consumedGas": 0,
+                        "invocationStack":
+                        [
+                            {
+                                "scriptHash": "W2jhyKUs0WrgEZhOJr5tY35y7NU=",
+                                "instructionPointer": 18,
+                                "nextInstruction": 0x66,
```
Comment (osmirnov) [inline test/NeoSharp.VM.Test/Tests/PUSHM1_to_PUSH16.json]: JSON does not support hex format.

File: test/NeoSharp.VM.Test/Tests/PUSHM1_to_PUSH16.json [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+                                "evaluationStack":
+                                [
+                                    {
+                                        "type": "Integer",
```
Comment (osmirnov) [inline test/NeoSharp.VM.Test/Tests/PUSHM1_to_PUSH16.json:46]: JSON is really verbose here. I think about micro syntax for the sake of clarity.
```
[
    "int 16",
    "int 15",
    "int 14",
    "int 13",
    "int 12",
    "int 11",
    "int 10",
    "int 9",
    "int 8",
    "int 7",
    "int 6",
    "int 5",
    "int 4",
    "int 3",
    "int 2",
    "int 1",
    "byte []",
    "int -1"
]
```

File: test/NeoSharp.VM.Test/Tests/PUSHM1_to_PUSH16.json [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+                                        "value": "16"
+                                    },
+                                    {
+                                        "type": "Integer",
+                                        "value": "15"
```
Comment (osmirnov) [inline test/NeoSharp.VM.Test/Tests/PUSHM1_to_PUSH16.json]: Can it be a real number not a string?

File: src/NeoSharp.Core/VM/StackAccessor.cs [hunk 4]
```diff
@@ -75,12 +81,26 @@ public byte[] PopByteArray()
...
+                    return interop?.Value;
+                }
             }
+            else
+            {
+                // Extract base type
+
+                if (typeof(T).IsAssignableFrom(stackItem.GetType()))
+                {
+                    return (T)(object)stackItem;
```
Comment (osmirnov) [inline src/NeoSharp.Core/VM/StackAccessor.cs:99]: `using`?

File: src/NeoSharp.VM/Extensions/InstructionParserExtensions.cs [hunk 5]
```diff
@@ -42,11 +42,30 @@ static InstructionParser()
...
+                return ms.ToArray();
+            }
+        }
+
         /// <summary>
         /// Parse script to instructions
         /// </summary>
         /// <param name="script">Script</param>
-        public IEnumerable<Instruction> Parse(byte[] script)
+        public static IEnumerable<Instruction> DecomileScript(this byte[] script)
```
Comment (osmirnov) [inline src/NeoSharp.VM/Extensions/InstructionParserExtensions.cs]: Typo: Decom**p**ile

File: src/NeoSharp.VM/Types/InteropService.cs [hunk 6]
```diff
@@ -61,17 +61,16 @@ public class InteropService
...
-
-            Register("System.ExecutionEngine.GetScriptContainer", GetScriptContainer);
-            Register("System.ExecutionEngine.GetExecutingScriptHash", GetExecutingScriptHash);
-            Register("System.ExecutionEngine.GetCallingScriptHash", GetCallingScriptHash);
-            Register("System.ExecutionEngine.GetEntryScriptHash", GetEntryScriptHash);
+            Register("Neo.Runtime.GetTrigger", NeoRuntimeGetTrigger, 1);
+            Register("Neo.Runtime.Log", NeoRuntimeLog, 1);
+            Register("Neo.Runtime.Notify", NeoRuntimeNotify, 1);
+            Register("Neo.Runtime.Serialize", "System.Runtime.Serialize", RuntimeSerialize, 1);
+            Register("Neo.Runtime.Deserialize", "System.Runtime.Deserialize", RuntimeDeserialize, 1);
```
Comment (osmirnov) [inline src/NeoSharp.VM/Types/InteropService.cs:68]: Iterators and Enumerators calls?

**Review by osmirnov — APPROVED** [review]

File: test/NeoSharp.VM.Test/Tests/PUSHM1_to_PUSH16.json (hunk 2 above)
Comment (shargon) [reply test/NeoSharp.VM.Test/Tests/PUSHM1_to_PUSH16.json:46]: I like it, could be easier to define like this
The problem is with structs and maps .. What is your idea ?

File: test/NeoSharp.VM.Test/Tests/PUSHM1_to_PUSH16.json (hunk 1 above)
Comment (shargon) [reply test/NeoSharp.VM.Test/Tests/PUSHM1_to_PUSH16.json]: I will use the enum (string) better

File: src/NeoSharp.Core/VM/StackAccessor.cs (hunk 4 above)
Comment (shargon) [reply src/NeoSharp.Core/VM/StackAccessor.cs:99]: I need this for take the type, and the base object (for compare with the json) so the using should be outside of this

File: test/NeoSharp.VM.Test/Tests/PUSHM1_to_PUSH16.json (hunk 2 above)
Comment (osmirnov) [reply test/NeoSharp.VM.Test/Tests/PUSHM1_to_PUSH16.json:46]: Maybe
```
    "struct [{int id: string Hello}]",
    "map [{int id: string World}]
```
I don't have all answers :D

Comment (shargon) [reply test/NeoSharp.VM.Test/Tests/PUSHM1_to_PUSH16.json:46]: xDD is hard to parse this kind of types

**Comment by shargon** [comment]
I will merge it until we found a better naming for `json`

File: test/NeoSharp.VM.Test/Tests/PUSHM1_to_PUSH16.json (hunk 2 above)
Comment (osmirnov) [reply test/NeoSharp.VM.Test/Tests/PUSHM1_to_PUSH16.json:46]: Then scrap this idea. It starts looking as yaml file 😉
